### PR TITLE
Cells

### PR DESCRIPTION
**Pull request by T-Nicholls**
All the non-unitconv changes I had stored up.

### CODE REVIEW

**Comment by T-Nicholls** [comment]
Also includes all the commits from #101, so I'll close that.

**Comment by willrogers** [comment]
Sorry, I didn't see this. I will look at it tomorrow morning.

File: pytac/data_source.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def set_value(self, field, value, throw):
 
 class DataSourceManager(object):
     """Class that manages all the data sources and UnitConv objects associated
-        with a lattice or element.
+    with a lattice or element.
 
     It recieves requests from a lattice or element object and directs them to
```
Comment (willrogers) [inline pytac/data_source.py]: receives (don't know where this came from originally)

File: pytac/lattice.py [hunk 2]
```diff
@@ -215,17 +259,18 @@ def get_length(self):
...
         return total_length
 
     def add_element(self, element):
-        """Append an element to the lattice.
+        """Append an element to the lattice and update its lattice reference.
 
         Args:
             element (Element): element to append.
         """
+        element._lattice = self
```
Comment (willrogers) [inline pytac/lattice.py]: This is setting to a private attribute of the element. We should have a `set_lattice()` or `add_to_lattice()` method.

File: pytac/lattice.py [hunk 3]
```diff
@@ -435,17 +480,32 @@ class EpicsLattice(Lattice):
...
+    .. Private Attributes:
+           _elements (list): The list of all the element objects in the lattice
+           _cs (ControlSystem): The control system to use for the more
+                                 efficient batch getting and setting of PVs.
+           _data_source_manager (DataSourceManager): A class that manages the
+                                                      data sources associated
+                                                      with this lattice.
     """
-    def __init__(self, name, epics_cs):
+    def __init__(self, name, epics_cs, symmetry=6):
```
Comment (willrogers) [inline pytac/lattice.py]: I don't think we should set a default symmetry.

File: test/test_element.py [hunk 4]
```diff
@@ -40,6 +70,14 @@ def test_get_unitconv_returns_unitconv_object(simple_element, unit_uc,
     assert simple_element.get_unitconv('y') == double_uc
 
 
+def tes_set_unit_conv(simple_element):
```
Comment (willrogers) [inline test/test_element.py]: `test` not `tes`

File: pytac/lattice.py [hunk 5]
```diff
@@ -489,42 +549,92 @@ def get_element_pv_names(self, family, field, handle):
...
-        values = self._cs.get_multiple(pv_names)
+        if data_source == pytac.DEFAULT:
+            data_source = self.get_default_data_source()
+        if units == pytac.DEFAULT:
+            units = self.get_default_units()
+        if data_source == pytac.LIVE:
+            pv_names = self.get_element_pv_names(family, field, handle)
+            values = self._cs.get_multiple(pv_names)
+            if units == pytac.PHYS:
+                converted_values = []
```
Comment (willrogers) [inline pytac/lattice.py]: Should we have a `convert_family_values()` method on `Lattice`?

File: pytac/lattice.py [hunk 6]
```diff
@@ -489,42 +549,92 @@ def get_element_pv_names(self, family, field, handle):
...
+                values = converted_values
+            pv_names = self.get_element_pv_names(family, field, pytac.SP)
+            if len(pv_names) != len(values):
+                raise IndexError("Number of elements in given array({0}) must "
+                                 "be equal to the number of elements in the "
+                                 "family({1}).".format(len(values),
+                                                       len(pv_names)))
+            self._cs.set_multiple(pv_names, values)
+        else:
+            elements = self.get_elements(family)
```
Comment (willrogers) [inline pytac/lattice.py]: Can't we use `Lattice.set_element_values()` here?

File: pytac/lattice.py [hunk 7]
```diff
@@ -489,42 +549,92 @@ def get_element_pv_names(self, family, field, handle):
...
+        if units == pytac.DEFAULT:
+            units = self.get_default_units()
+        if handle != pytac.SP:
+            raise HandleException("Must write using {0}.".format(pytac.SP))
+        if data_source == pytac.LIVE:
+            if units == pytac.PHYS:
+                converted_values = []
+                for elem, value in zip(self.get_elements(family), values):
+                    uc = elem.get_unitconv(field)
+                    converted_values.append(uc.convert(value,
```
Comment (willrogers) [inline pytac/lattice.py]: We could use `convert_family_values()` again here.

File: pytac/lattice.py (hunk 3 above)
Comment (T-Nicholls) [reply pytac/lattice.py]: added
```
    def set_lattice(self, lattice):
        """Set the stored lattice reference for this element to the passed
        lattice object.

        Args:
            lattice (Lattice): lattice object to store a reference to.
        """
        self._lattice = lattice
```

File: pytac/load_csv.py [hunk 8]
```diff
@@ -100,28 +100,33 @@ def load_unitconv(directory, mode, lattice):
...
+                    uc = unitconvs[int(item['uc_id'])]
                     if element.families.intersection(('HSTR', 'VSTR', 'QUAD',
                                                       'SEXT')):
-                        unitconvs[int(item['uc_id'])]._post_eng_to_phys = utils.get_div_rigidity(lattice.get_value('energy', units=pytac.PHYS))
-                        unitconvs[int(item['uc_id'])]._pre_phys_to_eng = utils.get_mult_rigidity(lattice.get_value('energy', units=pytac.PHYS))
-                    element._data_source_manager._uc[item['field']] = unitconvs[int(item['uc_id'])]
-                    element._data_source_manager._uc[item['field']].phys_units = item['phys_units']
-                    element._data_source_manager._uc[item['field']].eng_units = item['eng_units']
+                        energy = lattice.get_value('energy', units=pytac.PHYS)
+                        uc._post_eng_to_phys = utils.get_div_rigidity(energy)
```
Comment (willrogers) [inline pytac/load_csv.py]: Two more private attributes set here. Logically we add `set_post_eng_to_phys()` and `set_post_phys_to_eng()`.

File: pytac/lattice.py (hunk 2 above)
Comment (T-Nicholls) [reply pytac/lattice.py]: added
```
    def set_lattice(self, lattice):
        """Set the stored lattice reference for this element to the passed
        lattice object.

        Args:
            lattice (Lattice): lattice object to store a reference to.
        """
        self._lattice = lattice
```

**Pull request merged**